### ISSUE TITLE
14.0.1 does not build with LLVM_ENABLE_RUNTIMES if 14.0.0 is installed

### ISSUE THREAD

**Issue by q66** [llvm/llvm-project#54905]
When trying to build 14.0.1 with 14.0.0 already in the system, it will attempt to re-link some of the runtime stuff during install step, e.g. libc++. This will fail:

```
[ 92%] Linking CXX shared library /builddir/llvm-14.0.1/build/lib/libc++.so
cd /builddir/llvm-14.0.1/build/runtimes/runtimes-bins/libcxx/src && /usr/bin/cmake -E cmake_link_script CMakeFiles/cxx_shared.dir/link.txt --verbose=1
/builddir/llvm-14.0.1/build/./bin/clang++ --target=powerpc64le-chimera-linux-musl -fPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fstack-clash-protection -mtune=power9 -O2 -fdebug-prefix-map=/builddir/llvm-14.0.1=. -flto=thin -fPIC -g1 -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -ffunction-sections -fdata-sections --target=powerpc64le-chimera-linux-musl -O3  --target=powerpc64le-chimera-linux-musl -Wl,--as-needed -Wl,-z,now -Wl,-z,relro -flto=thin -Wl,-z,defs -Wl,-z,nodelete -fuse-ld=lld -shared -Wl,-soname,libc++.so.1 -o /builddir/llvm-14.0.1/build/lib/libc++.so.1.0 CMakeFiles/cxx_shared.dir/algorithm.cpp.o CMakeFiles/cxx_shared.dir/any.cpp.o CMakeFiles/cxx_shared.dir/atomic.cpp.o CMakeFiles/cxx_shared.dir/barrier.cpp.o CMakeFiles/cxx_shared.dir/bind.cpp.o CMakeFiles/cxx_shared.dir/charconv.cpp.o CMakeFiles/cxx_shared.dir/chrono.cpp.o CMakeFiles/cxx_shared.dir/condition_variable.cpp.o CMakeFiles/cxx_shared.dir/condition_variable_destructor.cpp.o CMakeFiles/cxx_shared.dir/exception.cpp.o CMakeFiles/cxx_shared.dir/functional.cpp.o CMakeFiles/cxx_shared.dir/future.cpp.o CMakeFiles/cxx_shared.dir/hash.cpp.o CMakeFiles/cxx_shared.dir/legacy_pointer_safety.cpp.o CMakeFiles/cxx_shared.dir/memory.cpp.o CMakeFiles/cxx_shared.dir/mutex.cpp.o CMakeFiles/cxx_shared.dir/mutex_destructor.cpp.o CMakeFiles/cxx_shared.dir/new.cpp.o CMakeFiles/cxx_shared.dir/optional.cpp.o CMakeFiles/cxx_shared.dir/random_shuffle.cpp.o CMakeFiles/cxx_shared.dir/ryu/d2fixed.cpp.o CMakeFiles/cxx_shared.dir/ryu/d2s.cpp.o CMakeFiles/cxx_shared.dir/ryu/f2s.cpp.o CMakeFiles/cxx_shared.dir/shared_mutex.cpp.o CMakeFiles/cxx_shared.dir/stdexcept.cpp.o CMakeFiles/cxx_shared.dir/string.cpp.o CMakeFiles/cxx_shared.dir/system_error.cpp.o CMakeFiles/cxx_shared.dir/thread.cpp.o CMakeFiles/cxx_shared.dir/typeinfo.cpp.o CMakeFiles/cxx_shared.dir/utility.cpp.o CMakeFiles/cxx_shared.dir/valarray.cpp.o CMakeFiles/cxx_shared.dir/variant.cpp.o CMakeFiles/cxx_shared.dir/vector.cpp.o CMakeFiles/cxx_shared.dir/debug.cpp.o CMakeFiles/cxx_shared.dir/random.cpp.o CMakeFiles/cxx_shared.dir/ios.cpp.o CMakeFiles/cxx_shared.dir/ios.instantiations.cpp.o CMakeFiles/cxx_shared.dir/iostream.cpp.o CMakeFiles/cxx_shared.dir/locale.cpp.o CMakeFiles/cxx_shared.dir/regex.cpp.o CMakeFiles/cxx_shared.dir/strstream.cpp.o CMakeFiles/cxx_shared.dir/filesystem/operations.cpp.o CMakeFiles/cxx_shared.dir/filesystem/directory_iterator.cpp.o  -nostdlib++ --unwindlib=none -lpthread -lc -lm -lrt /builddir/llvm-14.0.1/build/lib/clang/14.0.1/lib/powerpc64le-chimera-linux-musl/libclang_rt.builtins.a -fno-lto -rtlib=compiler-rt -Wl,--whole-archive,-Bstatic /builddir/llvm-14.0.1/build/lib/libc++abi.a -Wl,-Bdynamic,--no-whole-archive -lpthread -lc 
ld.lld: error: cannot open crtbeginS.o: No such file or directory
ld.lld: error: cannot open /builddir/llvm-14.0.1/build/lib/clang/14.0.0/lib/linux/libclang_rt.builtins-powerpc64le.a: No such file or directory
ld.lld: error: cannot open /builddir/llvm-14.0.1/build/lib/clang/14.0.0/lib/linux/libclang_rt.builtins-powerpc64le.a: No such file or directory
```

This is because even though it invokes the right clang binary, the `libclang` it uses is actually wrong, from the system. Therefore, it think it's actually clang 14.0.0, and will look for runtime in the wrong place. This does not happen during build, I assume because rpath is used.

**Comment by q66** [comment]
This can be mitigated with the following patch:

```
--- a/llvm/cmake/modules/LLVMExternalProjectUtils.cmake
+++ b/llvm/cmake/modules/LLVMExternalProjectUtils.cmake
@@ -321,7 +321,7 @@ function(llvm_ExternalProject_Add name source_dir)
                ${PASSTHROUGH_VARIABLES}
     INSTALL_COMMAND ""
     STEP_TARGETS configure build
-    BUILD_ALWAYS 1
+    BUILD_ALWAYS 0
     USES_TERMINAL_CONFIGURE 1
     USES_TERMINAL_BUILD 1
     USES_TERMINAL_INSTALL 1
```

That is more of a workaround than anything else, though.

**Comment by MaskRay** [comment]
This needs complete cmake command lines for others to reproduce.

For your `/builddir/llvm-14.0.1/build/./bin/clang++` command, you may add `-###  |& sed -E 's/ "?-[iIL]/\n&/g'` to dump the library paths.

**Comment by q66** [comment]
I ended up accidentally figuring this out.

I noticed that when I compiled a version of LLVM with a custom rpath and executed ldd on the resulting clang in the install environment, it was still pointing to `/usr/lib`. So I dug around... and realized that the install step in my build system is actually run through fakeroot. Which, internally, forces its own `LD_LIBRARY_PATH`, which has `/usr/lib`, and takes preference over builtin rpath. Which, subsequently, makes it use `/usr/lib/libclang.so.N` instead of the in-tree one, but only during the install step.

Apologies for the noise, closing.
